src/fields/field_python.impl.hh

```diff
@@ -186,34 +191,91 @@ void FieldPython<spacedim, Value>::value_list (const Armor::array &point_list, c
 template <int spacedim, class Value>
 void FieldPython<spacedim, Value>::set_value(FMT_UNUSED const Point &p, FMT_UNUSED const ElementAccessor<spacedim> &elm, FMT_UNUSED Value &value)
 {
-#ifdef FLOW123D_HAVE_PYTHON
-    for(unsigned int i = 0; i < spacedim; i++) {
-        p_value_ = PyFloat_FromDouble( p[i] );
-        PyTuple_SetItem(p_args_, i, p_value_);
-    }
-    p_value_ = PyObject_CallObject(p_func_, p_args_);
-    PythonLoader::check_error();
+    p_value_ = p_func_(p[0], p[1], p[2]);
 
     unsigned int pos =0;
     for(unsigned int row=0; row < value.n_rows(); row++)
         for(unsigned int col=0; col < value.n_cols(); col++, pos++)
-            if ( std::is_integral< typename Value::element_type >::value ) value(row,col) = PyLong_AsLong( PyTuple_GetItem( p_value_, pos ) );
-            else value(row,col) = PyFloat_AsDouble( PyTuple_GetItem( p_value_, pos ) );
+            if ( std::is_integral< typename Value::element_type >::value ) value(row,col) = p_value_[pos].cast<int>();
+            else value(row,col) = p_value_[pos].cast<double>();
+}
+
+
 
-#endif // FLOW123D_HAVE_PYTHON
+
+template <int spacedim, class Value>
+std::vector<const FieldCommon * > FieldPython<spacedim, Value>::set_dependency(FieldSet &field_set) {
+	std::vector<const FieldCommon *> required_fields;
+    py::list field_list;
+
+    try {
+        p_func_ = p_class_.attr("used_fields");
+        field_list = p_func_(p_class_);
+    } catch (const py::error_already_set &ex) {
+        PythonLoader::throw_error(ex);
+    }
+
+    std::vector<FieldCacheProxy> field_data;
+    std::vector<ssize_t> field_shape(2);
+
+    for (auto f : field_list) {
+        std::string field_name = f.cast<std::string>();
+        auto field_ptr = field_set.field(field_name);
+        if (field_ptr != nullptr) required_fields.push_back( field_ptr );
+        else {
+            field_ptr = field_set.user_field(field_name, this->time_);
+            if (field_ptr != nullptr) required_fields.push_back( field_ptr );
+            else THROW( ExcUnknownField() << EI_Field(field_name) << Input::EI_Address( in_rec_.address_string() ) );
+        }
+        field_shape[0] = ssize_t(field_ptr->shape_[0]);
+        field_shape[1] = ssize_t(field_ptr->shape_[1]);
+        field_data.emplace_back(field_name, field_shape, field_ptr->value_cache()->data_);
+    }
+
+    auto self_ptr = field_set.field(this->field_name_); // instance of FieldCommon of this field
+    std::vector<ssize_t> result_shape = { Value::NRows_, Value::NCols_ };
+    FieldCacheProxy result_data(this->field_name_, result_shape, self_ptr->value_cache()->data_);
+    p_func_ = p_class_.attr("set_dict");
+    py::list field_data_list = py::cast(field_data);
+    p_func_(p_class_, field_data_list, result_data);
```

Comment:
Program throws error on this line:

``
## 00:00:02.582 [0]  ERROR.TypeError: set_dict(): incompatible function arguments. The following argument types are supported:
##                       1. (self: flowpy.PythonFieldBaseCPP, arg0: List[FieldCacheProxy], arg1: FieldCacheProxy) -> None
##
##                   Invoked with: <class 'field_python_test.PyFieldWaterSourceDensity'>, [<flowpy.FieldCacheProxy object at 0x7f90f59f9330>], <flowpy.FieldCacheProxy object at 0x7f90f59fe9b0>
``

It is not clear to me which parameteris wrong. It is possible to reproduce the error in test 02 / 18